ranger/config/commands.py

```diff
@@ -1176,43 +1177,203 @@ def execute(self):
-    """
+
+    The symbolic mode is
+        symbolic mode   ::=  clause [, clause ...]
+        clause          ::=  [who ...] [action ...] action
+        action          ::=  op permcopy | op [perm ...]
+        who             ::=  u | g | o | a
+        op              ::=  = | + | -
+        permcopy        ::=  u | g | o
+        perm            ::=  r | w | x | s | t | X
+    """
+
+    WHO_RE = re.compile(r'''([ugoa]*)''', re.X)
+    ACTION_RE = re.compile(r'''
+        ([=+-])             # exactly one operator  (group 1)
+        (                   # perms                 (group 2)
+          [ugo]             #   either permcopy
+          |                 #   or
+          [rwxstX]*         #   permissionlist (may be empty)
+        )''', re.X)
+    OCTAL_MODE_RE = re.compile(r'''[0-7]+$''', re.X)
+    MASKS = {
+        'u': {'r': stat.S_IRUSR, 'w': stat.S_IWUSR, 'x': stat.S_IXUSR,
+              's': stat.S_ISUID, 't': 0},
+        'g': {'r': stat.S_IRGRP, 'w': stat.S_IWGRP, 'x': stat.S_IXGRP,
+              's': stat.S_ISGID, 't': 0},
+        'o': {'r': stat.S_IROTH, 'w': stat.S_IWOTH, 'x': stat.S_IXOTH,
+              's': 0,            't': stat.S_ISVTX},
+        'a': {},}
+
+    for perm in 'rwxst':
+        MASKS['a'][perm] = MASKS['u'][perm] | MASKS['g'][perm] | MASKS['o'][perm]
+
+    for entity in 'ugoa':
+        MASKS[entity]['rwx'] = (
+            MASKS[entity]['r'] | MASKS[entity]['w'] | MASKS[entity]['x'])
+        MASKS[entity]['rwxst'] = (MASKS[entity]['rwx'] |
+            MASKS[entity]['s'] | MASKS[entity]['t'])
+
+    def __init__(self, *args, **kwargs):
+        '''init chmod-command.'''
+        super(chmod, self).__init__(*args, **kwargs) # pylint: disable=R1725
+
+    def get_umask(self):
+        '''get os-umask.'''
+        result = os.umask(0o077)
+        os.umask(result)
+        return result
+
+    def fg_extract_rwx_for_source(self, source):
+        '''
+        return function which extracts for given entity ('u', 'g' or 'o')
+        the rwx-permissions as perm-string (e.g. 'rw') for the (original) mode.
+        '''
+        r_flag, w_flag = self.MASKS[source]['r'], self.MASKS[source]['w']
+        x_flag = self.MASKS[source]['x']
+        return lambda omode: (
+            ('r' if omode & r_flag else '') +
+            ('w' if omode & w_flag else '') +
+            ('x' if omode & x_flag else ''))
+
+    def fg_get_mode_for_target(self, target):
+        '''
+        return function which generatas bitmask for target
+        for the rwx-permissions (e.g. 'rw') input.
+        '''
+        r_flag, w_flag = self.MASKS[target]['r'], self.MASKS[target]['w']
+        x_flag = self.MASKS[target]['x']
+        return lambda rwx_str: (
+            (r_flag if 'r' in rwx_str else 0) |
+            (w_flag if 'w' in rwx_str else 0) |
+            (x_flag if 'x' in rwx_str else 0))
+
+    def get_mode_change_funcs(self, who, operator, perms):
```

Comment:
I have to admit I don't have the brainpower to properly review the logic here right now. Nothing stood out to me though.
Would you be up to writing some tests to verify the behavior?